File: flutter/lib/core/repositories/village_repository.dart
import '../services/village_api_service.dart';
import '../utils/result.dart';

abstract class IVillageRepository {
  Future<Result<List<dynamic>>> getDistricts();
  Future<Result<List<dynamic>>> getTalukas(String districtCode);
  Future<Result<List<dynamic>>> getVillages(String districtCode, String talukaCode);
  Future<Result<String>> resolveGisCode(String districtCode, String talukaCode, String villageCode);
}

class VillageRepository implements IVillageRepository {
  final VillageApiService apiService;

  VillageRepository({required this.apiService});

  @override
  Future<Result<List<dynamic>>> getDistricts() async {
    try {
      final res = await apiService.fetchDistricts();
      return Result.success(res);
    } catch (e) {
      return Result.failure('Failed to fetch districts: $e');
    }
  }

  @override
  Future<Result<List<dynamic>>> getTalukas(String districtCode) async {
    try {
      final res = await apiService.fetchTalukas(districtCode);
      return Result.success(res);
    } catch (e) {
      return Result.failure('Failed to fetch talukas: $e');
    }
  }

  @override
  Future<Result<List<dynamic>>> getVillages(String districtCode, String talukaCode) async {
    try {
      final res = await apiService.fetchVillages(districtCode, talukaCode);
      return Result.success(res);
    } catch (e) {
      return Result.failure('Failed to fetch villages: $e');
    }
  }

  @override
  Future<Result<String>> resolveGisCode(
      String districtCode, String talukaCode, String villageCode) async {
    try {
      final res = await apiService.resolveGisCode(districtCode, talukaCode, villageCode);
      return Result.success(res['gis_code'] as String? ?? 'MH-2701-270101-52001');
    } catch (e) {
      return Result.success('MH-$districtCode-$talukaCode-$villageCode');
    }
  }
}

File: backend/app/repositories/remote/village_repository.py
from typing import List
from app.providers.base_provider import LandRecordsProvider
from app.models.domain_models import District, Taluka, Village


class VillageRepository:
    def __init__(self, provider: LandRecordsProvider):
        self.provider = provider

    async def fetch_districts(self) -> List[District]:
        return await self.provider.get_districts()

    async def fetch_talukas(self, district_code: str) -> List[Taluka]:
        return await self.provider.get_talukas(district_code)

    async def fetch_villages(self, district_code: str, taluka_code: str) -> List[Village]:
        return await self.provider.get_villages(district_code, taluka_code)

    async def resolve_gis_code(
        self, district_code: str, taluka_code: str, village_code: str
    ) -> str:
        return await self.provider.get_village_gis_code(
            district_code, taluka_code, village_code
        )
